### PR TITLE
MAJOR: Improved the Hostnet ruleset

### PR DESCRIPTION
**Pull request by ping-localhost**
This is a backwards incompatible change for the fact that all `phpcs` commands will fail once this has been released.

The tool itself won't pass PHPCS once this has been merged into master, but I'll create a separate PR for that as to not pollute the current one.

---
**New sniff**

I've added a new sniff (`Hostnet.Classes.CopyrightMustBeBeforeDeclare`) that forces our copyright docblock to be defined before `declare(strict_types=1)`. The sniff is able to automatically fix it for the cases I've seen.

---
**Downsides:**

A few of the checks cannot be resolved automatically, which is too bad, I'll fix those by hand (yes really):
- `Generic.PHP.BacktickOperator`
- `Generic.Files.LineLength`
- `Squiz.Operators.ValidLogicalOperators.NotAllowed`
- `SlevomatCodingStandard.Classes.UnusedPrivateElements.UnusedProperty`

There might be a few more, but it's doesn't happen that often and it's quite easy to resolve.

--- 
**Benchmark:** 

Running `phpcs` with the current ruleset on one of our internal applications takes `1 mins, 27.05 secs`. With the changed ruleset the total time for `phpcs` was `1 mins, 14.48 secs`, however `phpcbf` took `3 mins, 53.41 secs` .

_EDIT_: This benchmark was done before the sniff `Hostnet.Classes.CopyrightMustBeBeforeDeclare` was added.

### CODE REVIEW

File: src/Hostnet/ruleset.xml [hunk 1]
```diff
@@ -1,73 +1,332 @@
...
+    <!-- Forbid empty statements -->
+    <rule ref="Generic.CodeAnalysis.EmptyStatement"/>
 
-    <!-- Have 8 chars padding maximum and always show as errors -->
-    <rule ref="Generic.Formatting.MultipleStatementAlignment">
+    <!-- Forbid final methods in final classes -->
+    <rule ref="Generic.CodeAnalysis.UnnecessaryFinalModifier"/>
+
+    <!-- Forbid useless empty method overrides -->
+    <rule ref="Generic.CodeAnalysis.UselessOverridingMethod"/>
```
Comment (linaori) [inline src/Hostnet/ruleset.xml:53]: Does this one take into account that the visibility _can_ change? I've seen this happen for a test once

File: src/Hostnet/ruleset.xml [hunk 2]
```diff
@@ -1,73 +1,332 @@
...
+        <type>error</type>
+    </rule>
+
+    <!-- Require comma after last element in multi-line array -->
+    <rule ref="SlevomatCodingStandard.Arrays.TrailingArrayComma"/>
+
+    <!-- Forbid dead code -->
+    <rule ref="SlevomatCodingStandard.Classes.UnusedPrivateElements">
+        <!-- Except private variables that are not used -->
+        <exclude name="SlevomatCodingStandard.Classes.UnusedPrivateElements.WriteOnlyProperty"/>
```
Comment (linaori) [inline src/Hostnet/ruleset.xml:100]: This is where you only write to them to keep a reference alive?

File: src/Hostnet/ruleset.xml [hunk 3]
```diff
@@ -1,73 +1,332 @@
...
+    <!-- Forbid comments starting with # -->
+    <rule ref="PEAR.Commenting.InlineComment"/>
+
+    <!-- Disallow else if in favor of elseif -->
+    <rule ref="PSR2.ControlStructures.ElseIfDeclaration.NotAllowed">
+        <type>error</type>
+    </rule>
+
+    <!-- Require comma after last element in multi-line array -->
+    <rule ref="SlevomatCodingStandard.Arrays.TrailingArrayComma"/>
```
Comment (linaori) [inline src/Hostnet/ruleset.xml:95]: THANK YOU!

File: src/Hostnet/ruleset.xml [hunk 4]
```diff
@@ -1,73 +1,332 @@
...
+
+    <!-- Forbid class being in a file with different name -->
     <rule ref="Squiz.Classes.ClassFileName"/>
+
+    <!-- Forbid duplicate properties -->
     <rule ref="Squiz.Classes.DuplicateProperty"/>
 
-    <exclude name="Hostnet.Commenting.FileCommentCopyright.MissingVersion"/>
+    <!-- Force `self::` for self-reference, force lower-case self, forbid spaces around `::` -->
+    <rule ref="Squiz.Classes.SelfMemberReference"/>
```
Comment (linaori) [inline src/Hostnet/ruleset.xml:256]: Be careful that this does not always work, how does this check treat scope changes in anonymous classes?

**Review by linaori — COMMENTED** [review]
Please double check some of my comments to see if they are not going to break anything

File: src/Hostnet/ruleset.xml (hunk 4 above)
Comment (ping-localhost) [reply src/Hostnet/ruleset.xml:256]: It takes anonymous classes into account as shown in it's tests:
- https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Standards/Squiz/Tests/Classes/SelfMemberReferenceUnitTest.inc
- https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Standards/Squiz/Tests/Classes/SelfMemberReferenceUnitTest.inc.fixed

File: src/Hostnet/ruleset.xml (hunk 2 above)
Comment (ping-localhost) [reply src/Hostnet/ruleset.xml:100]: For example, yes.

File: src/Hostnet/ruleset.xml (hunk 1 above)
Comment (linaori) [reply src/Hostnet/ruleset.xml:53]: This should be allowed:

```php
class A
{
    private function foo()
    {

    }
}

class B extends A
{
    protected function foo()
    {

    }
}

class C extends B
{
    public function foo()
    {

    }
}
```

Comment (ping-localhost) [reply src/Hostnet/ruleset.xml:53]: The sniff does not complain about this.

File: src/Hostnet/ruleset.xml (hunk 3 above)
Comment (ping-localhost) [reply src/Hostnet/ruleset.xml:95]: I've just updated the ruleset to work with the code snippet you sent me.

**Review by linaori — APPROVED** [review]
What about the readme?

**Comment by hboomsma** [comment]
* `SlevomatCodingStandard.Classes.UnusedPrivateElements.UnusedProperty` How does this behave with traits using the private properties (for example those generated by the `accessor-generator-plugin-lib`

**Comment by ping-localhost** [comment]
@iltar I've updated the [`README.md`](https://github.com/ping-localhost/phpcs-tool/blob/improve-ruleset/README.md) to be more slick and added a Wiki for all the rules we import and our custom sniffs.
@hboomsma It, indeed, doesn't work with the `accessor-generator-plugin-lib` so I've excluded it.

**Review by linaori — APPROVED** [review]
> It, indeed, doesn't work with the accessor-generator-plugin-lib so I've excluded it.

All the more reasons to not use traits :joy:

**Pull request merged**